Design/Crystallography/HKL.md

```diff
@@ -0,0 +1,121 @@
+
+```
+template <typename NumericType, typename Derived>
+bool isNonZero(const IsHKL<NumericType, Derived> &hkl) {
+  return std::all_of(hkl.cbegin(), hkl.cend(),
+                     [](const NumericType &index) { return index != NumericType{0}; });
+}
+```
+
+If functions perform transformations and return a transformed HKL, they must return a `Derived`-object:
+
+```
+template <typename NumericType, typename Derived>
+Derived higherOrderHKL(const IsHKL<NumericType, Derived> &hkl, const NumericType &order) {
+  Derived higherOrder;
+  std::transform(hkl.cbegin(), hkl.cend(), higherOrder.begin(),
+                 [=](const NumericType &index) { return index * order; });
+  return higherOrder;
+}
+```
+
+Inside these functions it is also always possible to use `static_cast` for turning the object into the derived type:
+
+```
+template <typename NumericType, typename Derived>
+NumericType indexSumOfDifference(const IsHKL<NumericType, Derived> &lhs, const IsHKL<NumericType, Derived> &rhs) {
+  Derived difference = static_cast<const Derived &>(lhs) - static_cast<const Derived &>(rhs);
+  
+  return std::accumulate(difference.cbegin(), difference.cend(), NumericType{0});
+}
+```
+
+
+### Suggested implementation
+
+There's a [PR](https://github.com/mantidproject/mantid/pull/15914) against the main repository where this design is implemented. It revealed one unfortunate name clash, `Mantid::Geometry::HKL` already exists as a sub-class of `MDFrame`. My suggestion for this is to use the opportunity and start migrating the crystallography related code into a new module called Crystallography, which then provides a new namespace `Mantid::Crystallography`. Initially I suggest that this new module sits between `Kernel` and `Geometry`. Once the existing crystallography-related code has been migrated, the dependency of the Geometry module on the Crystallography module can be removed.
+
+In the pull request there is a performance test as well that compares the performance for some common operations that are performed on HKLs between the different cases `int` and `double`, but also with `V3D`.
+
+Relative execution times (normed to `int`, lower is better):
+
+| Test   | `int`  | `double`  | `V3D`  |
+|---|---|---|---|
+| hkl1 == hkl2  | 1.0  | 1.6  | 2.3 |
+| hkl1 < hkl2   | 1.0  | 1.1  | 0.8 |
+| Matrix * hkl  | 1.0  | 1.5  | 4.8 |
+
+ Note that `V3D` does not do fuzzy comparison of floating point values for `operator<`, but it does for `operator==`. `IsHKL` does fuzzy comparison for both.
+
+Some of the difference is probably attributable to lack of inlining in `V3D`, but the main point is that the added type-safety doesn't result in runtime penalties.
+
+
+## Remaining issues
+
+Some open questions remain:
+
+  - Python export - not sure how well this transports through to Python
+  - Moving to new Crystallography-module
+  - Implement a "real world" example from the Geometry- or Crystal-module.
+
```

Comment:
One way would be to use `HKLGenerator` for constructing these types. I added that a few months ago, currently it lets you generate a vector of `V3D`s. It could be made a template (with default `V3D` so current code doesn't break) so you'd write something along these lines:

```
HKLGenerator<IntegerHKL> g(unitCell, resolution);

std::vector<IntegerHKL> hkls;
std::copy(g.cbegin(), g.cend(), hkls.begin());
```

Did you have any other use cases in mind?